qiskit/algorithms/minimum_eigen_solvers/qaoa/qaoa_circuit.py

```diff
@@ -0,0 +1,209 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2018, 2021.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Global X phases and parameterized problem hamiltonian."""
+
+from typing import Optional, Union, cast, Set
+
+import numpy as np
+
+from qiskit import QuantumCircuit
+from qiskit.circuit import Parameter, ParameterVector
+from qiskit.circuit.library import BlueprintCircuit
+from qiskit.opflow import OperatorBase, I, X, CircuitStateFn, H, CircuitOp, EvolutionFactory
+
+
+class QAOACircuit(BlueprintCircuit):
+    """Global X phases and parameterized problem hamiltonian."""
+
+    def __init__(self,
+                 cost_operator: OperatorBase,
+                 param_p: int,
+                 initial_state: Optional[QuantumCircuit] = None,
+                 mixer_operator: Optional[Union[QuantumCircuit, OperatorBase]] = None,
+                 name: str = "qaoa"):
```

Comment:
I see this is something to be passed to parent, which ends up being the name of the QuantumCircuit. I can see it might be useful to distinguish in someway one built from defaults, versus one build from a custom mixer. But I can see this can be confusing too if one names the default different things - unless you want identical circuits named differently for some reason. I see this is done this way in nlocal, but its subclasses like TwoLocal do not seem to override this (unless I missed it). If here would this be better by the classname as default - this way when printing the circuit where you can see the name, the association to the class is there more strongly.

Reply:
I don't a see a common pattern for naming sub-classes of `BlueprintCircuit` and `qaoa` as a name is good enough. A class name that comes from the internals of the implementation would not tell more to a user rather than a plain `qaoa` string.

@Cryoris how should we name circuits that sub-class `BlueprintCircuit`?

Reply:
Since class name was proposed to be QAOAAnsatz it seemed more precise to me than qaoa

Reply:
I'm good with "qaoa", since it's shorter and the whole QAOA algorithm essentially consists of applying this circuit 🙂

Reply:
I do prefer `qaoa` as well.